runtime/domain.c

```
@@ -16,79 +18,1442 @@
+  caml_sample_gc_collect(domain_state);
+  caml_remove_generational_global_root(&domain_state->dls_root);
+  caml_remove_generational_global_root(&domain_state->backtrace_last_exn);
+
+  caml_stat_free(domain_state->final_info);
+  caml_stat_free(domain_state->ephe_info);
+  caml_free_intern_state();
+  caml_free_extern_state();
+  caml_teardown_major_gc();
+  CAML_EVENTLOG_TEARDOWN();
+  caml_teardown_shared_heap(domain_state->shared_heap);
+  domain_state->shared_heap = 0;
+  caml_free_minor_tables(domain_state->minor_tables);
+  domain_state->minor_tables = 0;
+  caml_free_signal_stack();
+
+  if(domain_state->current_stack != NULL) {
+    caml_free_stack(domain_state->current_stack);
+  }
+
+  /* signal the domain termination to the backup thread
+     NB: for a program with no additional domains, the backup thread
+     will not have been started */
+  atomic_store_rel(&domain_self->backup_thread_msg, BT_TERMINATE);
+  caml_plat_signal(&domain_self->domain_cond);
+  caml_plat_unlock(&domain_self->domain_lock);
+
+  caml_plat_assert_all_locks_unlocked();
+  /* This is the last thing we do because we need to be able to rely
+     on caml_domain_alone (which uses caml_num_domains_running) in at least
+     the shared_heap lockfree fast paths */
+  atomic_fetch_add(&caml_num_domains_running, -1);
+}
+
+CAMLprim value caml_ml_domain_cpu_relax(value t)
+{
+  struct interruptor* self = &domain_self->interruptor;
+  handle_incoming_otherwise_relax (self);
+  return Val_unit;
+}
+
+CAMLprim value caml_domain_dls_set(value t)
+{
+  CAMLnoalloc;
+  caml_modify_generational_global_root(&Caml_state->dls_root, t);
+  return Val_unit;
+}
+
+CAMLprim value caml_domain_dls_get(value unused)
+{
+  CAMLnoalloc;
+  return Caml_state->dls_root;
+}
+
+CAMLprim value caml_ml_domain_set_name(value name)
+{
+  CAMLparam1(name);
+
+  if (caml_string_length(name) >= MAX_DOMAIN_NAME_LENGTH)
+    caml_invalid_argument("caml_ml_domain_set_name");
+  caml_thread_setname(String_val(name));
```

need to check the return value of `int caml_thread_setname()` here as it may return an error, so binding should raise exception in that case.

---

We could also view `thread_setname` as a best effort and ignore errors.  (The name is just to help with debugging and tracing, right?)

---

The C bindings currently signal success but do nothing if the platform doesn't implement a setname function, so ignoring the error would be the most consistent thing to do. I'll propose a PR for this after the merge (based on https://github.com/ocaml-multicore/ocaml-multicore/pull/829 but to ignore errors).

I think this might only become a potential security concern if we add a thread_get_name in the future and expose that to applications, but we don't right now.